src/xrpld/app/ledger/detail/InboundLedgers.cpp

```diff
@@ -75,18 +75,98 @@ class InboundLedgersImp : public InboundLedgers
         auto doAcquire = [&, seq, reason]() -> std::shared_ptr<Ledger const> {
             assert(hash.isNonZero());
 
-            // probably not the right rule
-            if (app_.getOPs().isNeedNetworkLedger() &&
-                (reason != InboundLedger::Reason::GENERIC) &&
-                (reason != InboundLedger::Reason::CONSENSUS))
+            bool const shouldAcquire = [&]() {
+                if (!app_.getOPs().isNeedNetworkLedger())
+                    return true;
+                if (reason == InboundLedger::Reason::GENERIC)
+                    return true;
+                if (reason == InboundLedger::Reason::CONSENSUS)
+                    return true;
+                return false;
+            }();
+            assert(
```

Comment:
This looks like unnecessary repetition of the condition used to initialise `shouldAcquire` but there is also a subtle bug here. We are reading `getOPs().isNeedNetworkLedger()` more than once, but this function returns an atomic (I mean its implementation in `NetworkOPs.cpp`) so in principle each time we call it, we might get a different response. I am not certain this condition can actually happen, but at the very least it is brittle code.

Also please note 3rd call to `getOPs().isNeedNetworkLedger()` inside logging below, the same applies as it could generate some confusing logs